tests/test_estimation/test_calculate_weights.py

```diff
@@ -0,0 +1,454 @@
+        [5, 14.0, "P", "211", "0001", 0, 2, False, 1.0],     # dropped by filter (formtype)
+        [6, 10.0, "P", "210", "0006", 0, 1, False, 6.3],
+        [7, 10.0, "P", "210", "0006", 1, 5, False, 1.0],     # dropped by filter (instance)
+        [8, np.nan, "P", "210", "0006", 1, 2, False, 4.0],   # dropped from calc as np.nan but weight applied
+        [9, 5.0, "P", "210", "0006", 0, 1, False, 6.3],
+        [10, 10.0, "P", "210", "0006", 0, 1, False, 6.3],
+    ]
+
+    expected_df = pd.DataFrame(data=data, columns=expected_cols)
+
+    expected_qa_cols = [
+        "cellnumber",
+        "N",
+        "n",
+        "outliers",
+        "a_weight",
+    ]
+
+    data = [
+        [1.0, 20.0, 4, 1.0, 6.3],
+        [2.0, 4.0, 1.0, 0.0, 4.0],
+        [4.0, 8.0, 1.0, 0.0, 8.0],
+    ]
+
+    expected_qa_df = pd.DataFrame(data=data, columns=expected_qa_cols)
+
+    result_df, result_qa_df = calw.calculate_weighting_factor(input_df, cellno_dict)
+
+    assert_frame_equal(result_df, expected_df, check_exact=False, rtol=0.01)
+    assert_frame_equal(result_qa_df, expected_qa_df, check_exact=False, rtol=0.01)
+
+
+def test_calculate_weighting_factor_709_with_missing():
+    """Test for calculate_weighting_factor for 709 numeric
+    with missing value"""
+
+    input_cols = [
+        "reference",
+        "709",
+        "selectiontype",
+        "statusencoded",
+        "formtype",
+        "instance",
+        "cellnumber",
+        "outlier",
+    ]
+
+    data = [
+        [1, '12', "P", "210", "0006", 0, 1, False],
+        [2, '14', "P", "211", "0006", 0, 2, False],
+        [2, 16, "P", "210", "0006", 0, 2, False],
+    ]
+
+    input_df = pd.DataFrame(data=data, columns=input_cols)
+
+    cellno_dict = {1: 20, 2: 4, 3: 6, 4: 8, 5: 10}
+
+    expected_cols = [
+        "reference",
+        "709",
+        "selectiontype",
+        "statusencoded",
+        "formtype",
+        "instance",
+        "cellnumber",
+        "outlier",
+        "a_weight",
+    ]
+
+    data = [
+        [1, 12, "P", "210", "0006", 0, 1, False, 6.3],
+        [2, 14, "P", "211", "0006", 0, 2, False, 4.0],
+        [2, 16, "P", "210", "0006", 0, 2, False, 4.0],
+    ]
+
+    expected_df = pd.DataFrame(data=data, columns=expected_cols)
+
+    result_df, result_qa_df = calw.calculate_weighting_factor(input_df, cellno_dict)
+
+    assert_series_equal(result_df["709"], expected_df["709"], check_dtype=True)
+
+
+def test_calculate_weighting_factor_709_no_missing():
+    """Test for calculate_weighting_factor for 709 numeric
+    without missing value"""
+
+    input_cols = [
+        "reference",
+        "709",
+        "selectiontype",
+        "statusencoded",
+        "formtype",
+        "instance",
+        "cellnumber",
+        "outlier",
+    ]
+
+    data = [
+        [1, '12', "P", "210", "0006", 0, 1, False],
+        [2, np.nan, "P", "211", "0006", 0, 2, False],
+        [2, 16, "P", "210", "0006", 0, 2, False],
```

Comment:
It appears that this and the above function names should be corrected since this has an np.nan while the above doesn't.
Happy to be corrected on this since I might be missing something

Reply:
Good spot!  I've mixed the names